sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/CPDTokensCache.kt

```diff
@@ -0,0 +1,165 @@
+    }
+
+    fun deserialize(serialized: ByteArray): List<CPDToken> {
+        return serialized
+            .decodeToString()
+            .split(DELIMITER)
+            .mapNotNull { stringToToken(it) }
+    }
+
+    fun stringToToken(serialized: String): CPDToken? {
+        if (serialized.isBlank()) {
+            return null
+        }
+        var startIndex = 0
+        var endIndex = serialized.indexOf(',')
+        try {
+            val startLine = serialized.slice(startIndex..endIndex - 1).trim().toInt()
+
+            startIndex = endIndex + 1
+            endIndex = serialized.indexOf(',', startIndex)
+            val startLineOffset = serialized.slice(startIndex..endIndex - 1).trim().toInt()
+
+            startIndex = endIndex + 1
+            endIndex = serialized.indexOf(',', startIndex)
+            val endLine = serialized.slice(startIndex..endIndex - 1).trim().toInt()
+
+            startIndex = endIndex + 1
+            endIndex = serialized.indexOf(',', startIndex)
+            val endLineOffset = serialized.slice(startIndex..endIndex - 1).trim().toInt()
+
+            startIndex = endIndex + 1
+            val range: TextRange = InternalTextRange(startLine, startLineOffset, endLine, endLineOffset)
+            val token = serialized.slice(startIndex..serialized.length - 1)
+
+            return CPDToken(range, token)
+        } catch (exception: NumberFormatException) {
+            return null
+        }
+
+
+    }
+}
+
+class InternalTextRange(
+    private val startLine: Int,
+    private val startLineOffset: Int,
+    private val endLine: Int,
+    private val endLineOffset: Int
+) : TextRange {
+    override fun start(): TextPointer {
+        return InternalTextPointer(startLine, startLineOffset)
+    }
+
+    override fun end(): TextPointer {
+        return InternalTextPointer(endLine, endLineOffset)
+    }
+
+    override fun overlap(another: TextRange): Boolean {
+        return this.end().compareTo(another.start()) > 0 &&
+            another.end().compareTo(this.start()) > 0
+    }
+
+    override fun equals(other: Any?): Boolean = other != null &&
+        other is TextRange &&
+        this.start().line().equals(other.start().line()) &&
+        this.start().lineOffset().equals(other.start().lineOffset()) &&
+        this.end().line().equals(other.end().line()) &&
+        this.end().lineOffset().equals(other.end().lineOffset())
+
+    override fun hashCode(): Int {
+        return Objects.hash(this.startLine, this.startLineOffset, this.endLine, this.endLineOffset)
+    }
+}
+
+class InternalTextPointer(
+    private val line: Int, private val lineOffset: Int
+) : TextPointer {
+    override fun compareTo(other: TextPointer?): Int {
+        return if (this.line == other!!.line()) {
+            Integer.compare(lineOffset, other.lineOffset())
+        } else Integer.compare(this.line, other.line())
+    }
+
+    override fun line() = line
+
+    override fun lineOffset() = lineOffset
+}
```

Comment:
Later EDIT: I think we don't actually need to implement TextRange ourselves. We have an InputFile available and should use it's API instead. Let's discuss more.